File: typescript/scheme/scheme_vibrant_test.ts
/**
 * @license
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'jasmine';

import {MaterialDynamicColors} from '../dynamiccolor/material_dynamic_colors.js';
import {Hct} from '../hct/hct.js';
import {customMatchers} from '../utils/test_utils.js';

import {SchemeVibrant} from './scheme_vibrant.js';

beforeEach(() => {
  jasmine.addMatchers(customMatchers);
});

describe('scheme vibrant test', () => {
  it('keyColors', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, 0.0);

    expect(MaterialDynamicColors.primaryPaletteKeyColor.getArgb(scheme)).matchesColor(0xff080CFF);
    expect(MaterialDynamicColors.secondaryPaletteKeyColor.getArgb(scheme)).matchesColor(0xff7B7296);
    expect(MaterialDynamicColors.tertiaryPaletteKeyColor.getArgb(scheme)).matchesColor(0xff886C9D);
    expect(MaterialDynamicColors.neutralPaletteKeyColor.getArgb(scheme)).matchesColor(0xff777682);
    expect(MaterialDynamicColors.neutralVariantPaletteKeyColor.getArgb(scheme)).matchesColor(0xff767685);
  });
  
  it('lightTheme_minContrast_primary', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, -1.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme)).matchesColor(0xff5660ff);
  });

  it('lightTheme_standardContrast_primary', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, 0.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme)).matchesColor(0xff343dff);
  });

  it('lightTheme_maxContrast_primary', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, 1.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme)).matchesColor(0xff000181);
  });

  it('lightTheme_minContrast_primaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, -1.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme)).matchesColor(0xffe0e0ff);
  });

  it('lightTheme_standardContrast_primaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, 0.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme)).matchesColor(0xffe0e0ff);
  });

  it('lightTheme_maxContrast_primaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, 1.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme)).matchesColor(0xff0000e3);
  });

  it('lightTheme_minContrast_onPrimaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, -1.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme)).matchesColor(0xff3e47ff);
  });

  it('lightTheme_standardContrast_onPrimaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, 0.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme)).matchesColor(0xff00006e);
  });

  it('lightTheme_maxContrast_onPrimaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, 1.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme)).matchesColor(0xffd6d6ff);
  });

  it('lightTheme_minContrast_surface', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, -1.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme)).matchesColor(0xfffbf8ff);
  });

  it('lightTheme_standardContrast_surface', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, 0.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme)).matchesColor(0xfffbf8ff);
  });

  it('lightTheme_maxContrast_surface', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), false, 1.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme)).matchesColor(0xfffbf8ff);
  });

  it('darkTheme_minContrast_primary', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, -1.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme)).matchesColor(0xff5660ff);
  });

  it('darkTheme_standardContrast_primary', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 0.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme)).matchesColor(0xffbec2ff);
  });

  it('darkTheme_maxContrast_primary', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 1.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme)).matchesColor(0xfff6f4ff);
  });

  it('darkTheme_minContrast_primaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, -1.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme)).matchesColor(0xff0000ef);
  });

  it('darkTheme_standardContrast_primaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 0.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme)).matchesColor(0xff0000ef);
  });

  it('darkTheme_maxContrast_primaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 1.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme)).matchesColor(0xffc4c6ff);
  });

  it('darkTheme_minContrast_onPrimaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, -1.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme)).matchesColor(0xffa9afff);
  });

  it('darkTheme_standardContrast_onPrimaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 0.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme)).matchesColor(0xffe0e0ff);
  });

  it('darkTheme_maxContrast_onPrimaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 1.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme)).matchesColor(0xff0001c6);
  });

  it('darkTheme_minContrast_onTertiaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, -1.0);
    expect(MaterialDynamicColors.onTertiaryContainer.getArgb(scheme)).matchesColor(0xffc9a9df);
  });

  it('darkTheme_standardContrast_onTertiaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 0.0);
    expect(MaterialDynamicColors.onTertiaryContainer.getArgb(scheme)).matchesColor(0xfff2daff);
  });

  it('darkTheme_maxContrast_onTertiaryContainer', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 1.0);
    expect(MaterialDynamicColors.onTertiaryContainer.getArgb(scheme)).matchesColor(0xff472e5b);
  });

  it('darkTheme_minContrast_surface', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, -1.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme)).matchesColor(0xff12131c);
  });

  it('darkTheme_standardContrast_surface', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 0.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme)).matchesColor(0xff12131c);
  });

  it('darkTheme_maxContrast_surface', () => {
    const scheme = new SchemeVibrant(Hct.fromInt(0xff0000ff), true, 1.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme)).matchesColor(0xff12131c);
  });
});

File: cpp/scheme/scheme_vibrant_test.cc
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cpp/scheme/scheme_vibrant.h"

#include "testing/base/public/gunit.h"
#include "cpp/cam/hct.h"
#include "cpp/dynamiccolor/material_dynamic_colors.h"

namespace material_color_utilities {

namespace {
TEST(SchemeVibrantTest, KeyColors) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryPaletteKeyColor().GetArgb(scheme),
            0xff080cff);
  EXPECT_EQ(MaterialDynamicColors::SecondaryPaletteKeyColor().GetArgb(scheme),
            0xff7b7296);
  EXPECT_EQ(MaterialDynamicColors::TertiaryPaletteKeyColor().GetArgb(scheme),
            0xff886c9d);
  EXPECT_EQ(MaterialDynamicColors::NeutralPaletteKeyColor().GetArgb(scheme),
            0xff777682);
  EXPECT_EQ(
      MaterialDynamicColors::NeutralVariantPaletteKeyColor().GetArgb(scheme),
      0xff767685);
}
TEST(SchemeVibrantTest, lightTheme_minContrast_primary) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, -1.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xff5660ff);
}
TEST(SchemeVibrantTest, lightTheme_standardContrast_primary) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xff343dff);
}
TEST(SchemeVibrantTest, lightTheme_maxContrast_primary) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, 1.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xff000181);
}
TEST(SchemeVibrantTest, lightTheme_minContrast_primaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, -1.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xffd5d6ff);
}
TEST(SchemeVibrantTest, lightTheme_standardContrast_primaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xffe0e0ff);
}
TEST(SchemeVibrantTest, lightTheme_maxContrast_primaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, 1.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xff0000e3);
}
TEST(SchemeVibrantTest, lightTheme_minContrast_onPrimaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, -1.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xff333dff);
}
TEST(SchemeVibrantTest, lightTheme_standardContrast_onPrimaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xff00006e);
}
TEST(SchemeVibrantTest, lightTheme_maxContrast_onPrimaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, 1.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xffffffff);
}
TEST(SchemeVibrantTest, lightTheme_minContrast_surface) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, -1.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xfffbf8ff);
}
TEST(SchemeVibrantTest, lightTheme_standardContrast_surface) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xfffbf8ff);
}
TEST(SchemeVibrantTest, lightTheme_maxContrast_surface) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), false, 1.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xfffbf8ff);
}
TEST(SchemeVibrantTest, darkTheme_minContrast_primary) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, -1.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xff7c84ff);
}
TEST(SchemeVibrantTest, darkTheme_standardContrast_primary) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 0.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xffbec2ff);
}
TEST(SchemeVibrantTest, darkTheme_maxContrast_primary) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 1.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xfffdf9ff);
}
TEST(SchemeVibrantTest, darkTheme_minContrast_primaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, -1.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xff0001c9);
}
TEST(SchemeVibrantTest, darkTheme_standardContrast_primaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 0.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xff0000ef);
}
TEST(SchemeVibrantTest, darkTheme_maxContrast_primaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 1.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xffc4c6ff);
}
TEST(SchemeVibrantTest, darkTheme_minContrast_onPrimaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, -1.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xff949bff);
}
TEST(SchemeVibrantTest, darkTheme_standardContrast_onPrimaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 0.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xffe0e0ff);
}
TEST(SchemeVibrantTest, darkTheme_maxContrast_onPrimaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 1.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xff000000);
}
TEST(SchemeVibrantTest, darkTheme_minContrast_onTertiaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, -1.0);
  EXPECT_EQ(MaterialDynamicColors::OnTertiaryContainer().GetArgb(scheme),
            0xffb798cd);
}
TEST(SchemeVibrantTest, darkTheme_standardContrast_onTertiaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 0.0);
  EXPECT_EQ(MaterialDynamicColors::OnTertiaryContainer().GetArgb(scheme),
            0xfff2daff);
}
TEST(SchemeVibrantTest, darkTheme_maxContrast_onTertiaryContainer) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 1.0);
  EXPECT_EQ(MaterialDynamicColors::OnTertiaryContainer().GetArgb(scheme),
            0xff000000);
}
TEST(SchemeVibrantTest, darkTheme_minContrast_surface) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, -1.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xff12131c);
}
TEST(SchemeVibrantTest, darkTheme_standardContrast_surface) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 0.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xff12131c);
}
TEST(SchemeVibrantTest, darkTheme_maxContrast_surface) {
  SchemeVibrant scheme = SchemeVibrant(Hct(0xff0000ff), true, 1.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xff12131c);
}
}  // namespace
}  // namespace material_color_utilities
